hive-metastore/src/test/java/org/apache/iceberg/hive/HiveMetastoreExtension.java

```diff
@@ -0,0 +1,84 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iceberg.hive;
+
+import java.util.Map;
+import org.apache.hadoop.hive.conf.HiveConf;
+import org.apache.hadoop.hive.metastore.HiveMetaStoreClient;
+import org.apache.hadoop.hive.metastore.api.Database;
+import org.apache.iceberg.relocated.com.google.common.collect.Maps;
+import org.junit.jupiter.api.extension.AfterEachCallback;
+import org.junit.jupiter.api.extension.BeforeEachCallback;
+import org.junit.jupiter.api.extension.ExtensionContext;
+
+public class HiveMetastoreExtension implements BeforeEachCallback, AfterEachCallback {
```

Comment:
Should this be `BeforeAllCallback` and `AfterEachCallback` ?

I don't think starting and stopping metastore for each test case is a good idea

Reply:
@ajantha-bhat , we are planning to extend [CatalogTests](https://github.com/apache/iceberg/blob/main/core/src/test/java/org/apache/iceberg/catalog/CatalogTests.java) for Hive too. It expects all the catalog related infos to be dropped after each tests.

Reply:
If we drop all namespaces and tables after each test, catalog tests can work fine IMO. 

I think starting and stopping metastore for each tests is still expensive. 

If others (@nastra) is ok with this change, Then fine.

Reply:
I wrongly remembered that we needed this for https://github.com/apache/iceberg/pull/8918, but in fact we can just init the catalog before each method in `TestHiveCatalog` for https://github.com/apache/iceberg/pull/8918, which means that we can indeed start/stop the metastore using the BeforeAllCallback/AfterAllCallback.

I have pushed a commit that addresses this

Reply:
Unless I am missing something using `BeforeAllCallback, AfterAllCallback`  with HiveMetastoreExtension will stop us using the `CatalogTests`. I mean it may ask many custom changes with `TestHiveCatalog`. 
Attaching a reference for the same. 
![Uploading Screenshot 2023-12-12 at 7.30.37 PM.png…]()

Reply:
@nk1506: 
we can add a proper teardown in the beforeEach() like Nessie does when we extend it.
https://github.com/apache/iceberg/blob/0c5b87a27e0564ddaae2fa1cb93c32dd7a5e8681/nessie/src/test/java/org/apache/iceberg/nessie/TestNessieCatalog.java#L97

I think Nessie server is up only once. Instead of per testcase.
https://github.com/projectnessie/nessie/blob/main/servers/jax-rs-testextension/src/main/java/org/projectnessie/jaxrs/ext/NessieJaxRsExtension.java

Reply:
With Hive it can't drop namespace unless all the tables have been dropped. It throws error like.
`InvalidOperationException(message:Database newdb is not empty. One or more tables exist.)`
I had to [revert](https://github.com/apache/iceberg/pull/8918/files#diff-17a52cf779b5dc58f09db91962cb69514957356c8fe74dc8df36d9dbec8074d7R30) and added it as teardown of metastore after every tests.